src/main/java/org/broadinstitute/barclay/argparser/TaggedArgumentParser.java

```diff
@@ -0,0 +1,333 @@
+            final String rawArgumentValue,  // the raw argument value provided by the user
+            final String tagString          // the tag string that has been peeled off of the raw option string, including logical name and any attributes
+    )
+    {
+        final String surrogateKey = makeSurrogateKey(rawOptionString, rawArgumentValue);
+        if (null != tagSurrogates.put(surrogateKey, Pair.of(tagString, rawArgumentValue))) {
+            throw new CommandLineException.BadArgumentValue(
+                    String.format("The argument value: \"%s %s\" was duplicated on the command line", rawOptionString, rawArgumentValue));
+        }
+        return surrogateKey;
+    }
+
+    /**
+     * Construct a surrogate key from a raw option string (as provided by the user) and raw argument value (as
+     * provided by the user). In order to ensure uniqueness of the key, it needs to include all elements from the
+     * command line, including:
+     *
+     *  -the option prefix (short/long) used
+     *  -the option name (short/long) used
+     *  -the tag string used
+     *  -the argument value used
+     *
+     * @param rawOptionString the entire raw option string as provided by the user
+     * @param rawArgumentValue the entire raw argument value provided by the user
+     * @return a surrogate key that can be included as a substitute value in the command line args presented to
+     * command line the parser
+     */
+    private String makeSurrogateKey(
+            final String rawOptionString,
+            final String rawArgumentValue
+    ) {
+        // The keys strings are never parsed, but the separator is added to the
+        // middle to make visual key inspection easier
+        return rawOptionString + ARGUMENT_TAG_NAME_SEPARATOR + rawArgumentValue;
+    }
+
+    /**
+     * Parse a tag string and populate a TaggedArgument with values.
+     * @param tagString tag string (including logical name and attributes, no option name)
+     * @param taggedArg TaggableArgument to receive tags
+     */
+    public void populateArgumentTags(final TaggedArgument taggedArg, final String tagString) {
+        final ParsedArgument pa = ParsedArgument.of(tagString);
+        taggedArg.setTag(pa.getName());
+        taggedArg.setTagAttributes(pa.keyValueMap());
+    }
+
+    /**
+     * Given a TaggedArgument implementation, return a string representation of the tags and attributes
+     * for display purposes. The string returned does not include the argument name.
+     *
+     * @param taggedArg implementation of TaggedArgument interface
+     * @return a display string representing the tag name and attributes
+     */
+    public static String getDisplayString(final TaggedArgument taggedArg) {
+        Utils.nonNull(taggedArg);
+
+        StringBuilder sb = new StringBuilder();
+        sb.append(ARGUMENT_TAG_NAME_SEPARATOR);
+        sb.append(taggedArg.getTag() == null ? "NoTagName" : taggedArg.getTag());
```

Comment:
If a tagged argument has no tag, its display string should be simply the argument name. The case `getTag() == null && getTagAttributes() != null` should be disallowed, so you don't have to worry about it here.